File: crates/prolly/bindings/python/examples/provenance_values.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from tempfile import TemporaryDirectory
import sys

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

import prolly


def b(value: str) -> bytes:
    return value.encode("utf-8")


def text(value: bytes | None) -> str | None:
    return None if value is None else value.decode("utf-8")


def assert_bytes(expected: bytes, actual: bytes | None, label: str) -> None:
    if actual != expected:
        raise AssertionError(f"{label}: expected {expected!r}, got {actual!r}")


def upsert(key: str, value: str | bytes) -> prolly.MutationRecord:
    payload = value if isinstance(value, bytes) else b(value)
    return prolly.MutationRecord(kind=prolly.MutationKind.UPSERT, key=b(key), value=payload)


def delete(key: str) -> prolly.MutationRecord:
    return prolly.MutationRecord(kind=prolly.MutationKind.DELETE, key=b(key), value=None)

def provenance_values() -> None:
    engine = prolly.ProllyEngine.memory(prolly.default_config())
    source = b"CrabDB language bindings design"
    source_cid = prolly.cid_from_bytes(source).hex()
    chunk_cid = prolly.cid_from_bytes(source[:16]).hex()

    tree = engine.batch(
        engine.create(),
        [
            upsert("provenance/chunk/file-1/chunk-1", f"source={source_cid}|chunk={chunk_cid}|parser=v1"),
            upsert("provenance/claim/file-1/claim-1", "CrabDB uses Rust-backed bindings|chunk=file-1/chunk-1"),
        ],
    )

    claims = engine.range(tree, b"provenance/claim/file-1/", b"provenance/claim/file-10")
    assert len(claims) == 1
    assert b"Rust-backed" in claims[0].value

    print("provenance_values: claim links back to source and chunk CIDs")


if __name__ == "__main__":
    provenance_values()

File: crates/prolly/bindings/ruby/examples/provenance_values.rb
# frozen_string_literal: true

require 'tmpdir'

$LOAD_PATH.unshift(File.expand_path('../lib', __dir__))
require 'prolly'

def upsert(key, value)
  Prolly::MutationRecord.new(kind: Prolly::MutationKind::UPSERT, key: key.b, value: value.b)
end

def delete(key)
  Prolly::MutationRecord.new(kind: Prolly::MutationKind::DELETE, key: key.b, value: nil)
end

def assert_equal(expected, actual, label = 'value')
  return if expected == actual

  raise "#{label}: expected #{expected.inspect}, got #{actual.inspect}"
end

def provenance_values
  engine = Prolly::ProllyEngine.memory(Prolly.default_config)
  source = 'CrabDB language bindings design'.b
  source_cid = Prolly.cid_from_bytes(source).unpack1('H*')
  chunk_cid = Prolly.cid_from_bytes(source[0, 16]).unpack1('H*')
  tree = engine.batch(
    engine.create,
    [
      upsert('provenance/chunk/file-1/chunk-1', "source=#{source_cid}|chunk=#{chunk_cid}|parser=v1"),
      upsert('provenance/claim/file-1/claim-1', 'CrabDB uses Rust-backed bindings|chunk=file-1/chunk-1')
    ]
  )

  claims = engine.range(tree, 'provenance/claim/file-1/'.b, 'provenance/claim/file-10'.b)
  assert_equal 1, claims.length, 'claim count'
  raise 'missing claim text' unless claims.first.value.include?('Rust-backed'.b)

  puts 'provenance_values: claim links back to source and chunk CIDs'
end

provenance_values
